extensions/amp-tiktok/0.1/amp-tiktok.js

```diff
@@ -0,0 +1,253 @@
+    );
+  }
+
+  /** @override */
+  buildCallback() {
+    const {src} = this.element.dataset;
+    const videoIdRegex = /^((.+\/)?)(\d+)\/?$/;
+    if (src) {
+      this.videoId_ = src.replace(videoIdRegex, (_, _1, _2, id) => id);
+    } else {
+      const blockquoteOrNull = childElementByTag(this.element, 'blockquote');
+      if (
+        !blockquoteOrNull ||
+        !blockquoteOrNull.hasAttribute('placeholder') ||
+        !blockquoteOrNull.dataset.videoId
+      ) {
+        return;
+      }
+      this.videoId_ = blockquoteOrNull.dataset.videoId;
+    }
+  }
+
+  /** @override */
+  layoutCallback() {
+    const {locale = 'en-US'} = this.element.dataset;
+    const src = `https://www.tiktok.com/embed/v2/${encodeURIComponent(
+      this.videoId_
+    )}?lang=${encodeURIComponent(locale)}`;
+
+    const iframe = createElementWithAttributes(
+      this.element.ownerDocument,
+      'iframe',
+      {
+        'src': src,
+        'name': this.iframeNameString_,
+        'aria-hidden': 'true',
+        'frameborder': '0',
+      }
+    );
+
+    iframe.classList.add('i-amphtml-tiktok-unresolved');
+    this.iframe_ = iframe;
+
+    this.unlistenMessage_ = listen(
+      this.win,
+      'message',
+      this.handleTiktokMessages_.bind(this)
+    );
+
+    this.element.appendChild(iframe);
+    return this.loadPromise(iframe).then(() => {
+      const {promise, resolve} = new Deferred();
+
+      this.resolveRecievedFirstMessage_ = resolve;
+      Services.timerFor(this.win)
+        .timeoutPromise(1000, promise)
+        .catch(() => {
+          // If no resize messages are recieved the fallback is to
+          // resize to '525px' in height.
+          this.resizeOuter_('525px');
+          setStyles(this.iframe_, {
+            'width': px('325px'),
+            'height': px('525px'),
+          });
+        });
+    });
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  handleTiktokMessages_(event) {
+    if (
+      event.origin != 'https://www.tiktok.com' ||
+      event.source != this.iframe_.contentWindow
+    ) {
+      return;
+    }
+    const data = tryParseJson(getData(event));
+    if (!data) {
+      return;
+    }
+    if (data['height']) {
+      if (this.resolveRecievedFirstMessage_) {
+        this.resolveRecievedFirstMessage_();
+      }
+      this.resizeOuter_(data['height']);
+      setStyles(this.iframe_, {
+        'width': px(data['width']),
+        'height': px(data['height']),
+      });
+    }
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+    return true; // layout again
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * Return unique name with the appropriate prefix.
+   * This name is defined by tiktok but is not documented on their site.
```

Comment:
Scary! Is this documented elsewhere? Tested somehow with e2e that makes it clear the API has changed shape?

Reply:
Also, does this component continue to operate if unable to parse the iframe name string?

Reply:
I don't believe its documented in TikTok's documentation. In a fast follow to this PR I am adding an e2e which will verify if this API is still working. 
The component will continue to operate through the resize fallback mechanism, however the contents will likely be cropped.

Reply:
After an offline conversation: it might be best to forgo messaging and render in one size only.